pkg/virt-controller/watch/pool.go

```diff
@@ -0,0 +1,980 @@
+	objs, err := c.vmInformer.GetIndexer().ByIndex(cache.NamespaceIndex, namespace)
+	if err != nil {
+		return nil, err
+	}
+	vms := []*virtv1.VirtualMachine{}
+	for _, obj := range objs {
+		vms = append(vms, obj.(*virtv1.VirtualMachine))
+	}
+	return vms, nil
+}
+
+func (c *PoolController) calcDiff(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine) int {
+	wantedReplicas := int32(1)
+	if pool.Spec.Replicas != nil {
+		wantedReplicas = *pool.Spec.Replicas
+	}
+
+	return len(vms) - int(wantedReplicas)
+}
+
+func filterDeletingVMs(vms []*virtv1.VirtualMachine) []*virtv1.VirtualMachine {
+
+	filtered := []*virtv1.VirtualMachine{}
+	for _, vm := range vms {
+		if vm.DeletionTimestamp == nil {
+			filtered = append(filtered, vm)
+		}
+	}
+	return filtered
+}
+
+func (c *PoolController) scaleIn(pool *poolv1.VirtualMachinePool, vms []*virtv1.VirtualMachine, count int) error {
+
+	poolKey, err := controller.KeyFunc(pool)
+	if err != nil {
+		return err
+	}
+
+	elgibleVMs := filterDeletingVMs(vms)
+
+	// make sure we count already deleting VMs here during scale in.
+	count = count - (len(vms) - len(elgibleVMs))
+
+	if len(elgibleVMs) == 0 || count == 0 {
+		return nil
+	} else if count > len(elgibleVMs) {
+		count = len(elgibleVMs)
+	}
+
+	// random delete strategy
+	rand.Shuffle(len(elgibleVMs), func(i, j int) {
+		elgibleVMs[i], elgibleVMs[j] = elgibleVMs[j], elgibleVMs[i]
+	})
+
+	log.Log.Object(pool).Infof("Removing %d VMs from pool", count)
+
+	var wg sync.WaitGroup
+
+	deleteList := elgibleVMs[0:count]
+	c.expectations.ExpectDeletions(poolKey, controller.VirtualMachineKeys(deleteList))
+	wg.Add(len(deleteList))
+	errChan := make(chan error, len(deleteList))
+	for i := 0; i < len(deleteList); i++ {
+		go func(idx int) {
+			defer wg.Done()
+			vm := deleteList[idx]
+
+			foreGround := metav1.DeletePropagationForeground
+			err := c.clientset.VirtualMachine(vm.Namespace).Delete(vm.Name, &metav1.DeleteOptions{PropagationPolicy: &foreGround})
+			if err != nil {
+				c.expectations.DeletionObserved(poolKey, controller.VirtualMachineKey(vm))
+				c.recorder.Eventf(pool, k8score.EventTypeWarning, FailedDeleteVirtualMachineReason, "Error deleting virtual machine %s: %v", vm.ObjectMeta.Name, err)
+				errChan <- err
+				return
+			}
+			c.recorder.Eventf(pool, k8score.EventTypeNormal, SuccessfulDeleteVirtualMachineReason, "Deleted VM %s/%s with uid %v from pool", vm.Namespace, vm.Name, vm.ObjectMeta.UID)
+			log.Log.Object(pool).Infof("Deleted vm %s/%s from pool", vm.Namespace, vm.Name)
+		}(i)
+	}
+
+	wg.Wait()
+
+	select {
+	case err := <-errChan:
+		// Only return the first error which occurred. We log the rest
+		return err
+	default:
+	}
+
+	return nil
+}
+
+func generateVMName(index int, baseName string) string {
+	return fmt.Sprintf("%s-%d", baseName, index)
+}
+
+func calculateNewVMNames(count int, baseName string, namespace string, vmStore cache.Store) []string {
```

Comment:
Only one vm pool is being processed by a thread at a single time, so calculateNewVMNames will only run for a single VM once in parallel. Two threads shouldn't be able to collide because the names are created with the vm pool's name as a prefix.

Basically, since two VM pools can't have the same name, and a pool is only ever processed by a single thread at a time, there isn't a possiblity for this kind of name collision.

Someone could create a random VM which collides wiht a vmpool name though, but in that case the calculateNewVMNames simply skips that vm name and doesn't count it towards the total.